qbraid_qir/qasm3/analyzer.py

```diff
@@ -0,0 +1,184 @@
+
+        """
+        indices = []
+        var_name = None
+        comma_separated = False
+
+        if isinstance(index_expr.collection, IndexExpression):
+            while isinstance(index_expr, IndexExpression):
+                indices.append(index_expr.index[0])
+                index_expr = index_expr.collection
+        else:
+            comma_separated = True
+            indices = index_expr.index
+
+        var_name = index_expr.collection.name if comma_separated else index_expr.name
+        if not comma_separated:
+            indices = indices[::-1]
+
+        return var_name, indices
+
+    @staticmethod
+    def find_array_element(multi_dim_arr: list[Any], indices: list[int]) -> Any:
+        """Find the value of an array at the specified indices.
+
+        Args:
+            multi_dim_arr (list): The multi-dimensional list to search.
+            indices (list[int]): The indices to search.
+
+        Returns:
+            Any: The value at the specified indices.
+        """
+        temp = multi_dim_arr
+        for index in indices:
+            temp = temp[index]
+        return temp
+
+    @staticmethod
+    def analyse_branch_condition(
+        condition: Union[UnaryExpression, BinaryExpression, IndexExpression]
+    ) -> bool:
+        """
+        analyze the branching condition to determine the branch to take
+
+        Args:
+            condition (Any): The condition to analyze
+
+        Returns:
+            bool: The branch to take
+        """
+
+        if isinstance(condition, UnaryExpression):
+            if condition.op.name != "!":
+                raise_qasm3_error(
+                    message=f"Unsupported unary expression '{condition.op.name}' in if condition",
+                    err_type=Qasm3ConversionError,
+                    span=condition.span,
+                )
+            return False
+        if isinstance(condition, BinaryExpression):
+            if condition.op.name != "==":
+                raise_qasm3_error(
+                    message=f"Unsupported binary expression '{condition.op.name}' in if condition",
+                    err_type=Qasm3ConversionError,
+                    span=condition.span,
+                )
+            if not isinstance(condition.lhs, IndexExpression):
+                raise_qasm3_error(
+                    message=f"Unsupported expression type '{type(condition.rhs)}' in if condition",
+                    err_type=Qasm3ConversionError,
+                    span=condition.span,
+                )
+            return condition.rhs.value != 0
+        if not isinstance(condition, IndexExpression):
+            raise_qasm3_error(
+                message=(
+                    f"Unsupported expression type '{type(condition)}' in if condition. "
+                    "Can only be a simple comparison"
+                ),
+                err_type=Qasm3ConversionError,
+                span=condition.span,
+            )
```

Comment:
Based on the type hint `condition: Union[UnaryExpression, BinaryExpression, IndexExpression]`, this code is unreachable.